File: sdks/typescript/test/client.test.ts
import { test, describe, after } from 'node:test';
import assert from 'node:assert/strict';
import { createServer, type Server, type IncomingMessage } from 'node:http';
import { AtlasClient, AtlasError, AtlasConnectionError } from '../src/index.js';

/**
 * Tests run against a real HTTP server on a real port, so the actual fetch
 * path, header handling, retry loop, and JSON parsing are all exercised. A
 * mocked fetch would test the mock.
 */

interface Recorded {
  method: string;
  url: string;
  headers: Record<string, string | string[] | undefined>;
  body: string;
}

let recorded: Recorded[] = [];
/** Queue of responses; each request pops one. Falls back to 200 {}. */
let responses: Array<{ status: number; body: string; contentType?: string }> = [];

const server: Server = createServer((req: IncomingMessage, res) => {
  const chunks: Buffer[] = [];
  req.on('data', (c: Buffer) => chunks.push(c));
  req.on('end', () => {
    recorded.push({
      method: req.method ?? '',
      url: req.url ?? '',
      headers: req.headers,
      body: Buffer.concat(chunks).toString(),
    });
    const next = responses.shift() ?? { status: 200, body: '{}' };
    res.writeHead(next.status, {
      'content-type': next.contentType ?? 'application/json',
    });
    res.end(next.body);
  });
});

// Started with top-level await rather than a root `before` hook.
//
// On the oldest Node this package supports (20.3), a root-level `before`
// does not reliably complete before the tests inside top-level
// `describe` blocks begin, so `baseUrl` was still undefined when the
// first client was constructed — the whole suite failed on 20.3 while
// passing on 22. Module evaluation, by contrast, is guaranteed to finish
// before any test runs on every supported version.
await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
const addr = server.address();
if (typeof addr === 'string' || addr === null) throw new Error('no port');
const baseUrl = `http://127.0.0.1:${addr.port}`;
// A listening socket would otherwise keep the process alive if `after`
// never runs; the suite should exit on its own either way.
server.unref();

after(async () => {
  await new Promise<void>((resolve, reject) =>
    server.close((e) => (e ? reject(e) : resolve())),
  );
});

function reset(): void {
  recorded = [];
  responses = [];
}

const TEST_PROJECT_KEY = 'atl_test_0123456789abcdef0123456789abcdef';

function client(opts: { token?: string; maxRetries?: number } = {}): AtlasClient {
  return new AtlasClient({
    baseUrl,
    projectKey: TEST_PROJECT_KEY,
    maxRetries: opts.maxRetries ?? 0,
    timeoutMs: 2_000,
    ...(opts.token ? { token: opts.token } : {}),
  });
}

describe('auth', () => {
  test('register maps snake_case to camelCase', async () => {
    reset();
    responses.push({ status: 201, body: '{"user_id":"u-1"}' });
    const out = await client().auth.register({
      email: 'a@b.dev',
      password: 'pw',
    });
    assert.equal(out.userId, 'u-1');
    assert.equal(recorded[0]?.url, '/v1/auth/register');
    assert.deepEqual(JSON.parse(recorded[0]!.body), {
      email: 'a@b.dev',
      password: 'pw',
    });
  });

  test('login stores the token for later calls', async () => {
    reset();
    responses.push({ status: 200, body: '{"token":"tok-1","expires_at":99}' });
    responses.push({
      status: 200,
      body: '{"user_id":"u-1","session_id":"s-1","last_lat":0,"last_lng":0,"issued_at":1,"expires_at":99}',
    });

    const c = client();
    const session = await c.auth.login({ email: 'a@b.dev', password: 'pw' });
    assert.equal(session.token, 'tok-1');
    assert.equal(session.expiresAt, 99);
    assert.equal(c.getToken(), 'tok-1');

    await c.auth.me();
    assert.equal(recorded[1]?.headers['authorization'], 'Bearer tok-1');
  });

  test('login omits coordinates unless both are given', async () => {
    reset();
    responses.push({ status: 200, body: '{"token":"t","expires_at":1}' });
    // The gateway 400s on one coordinate without the other, so a
    // half-filled form must not be forwarded as-is.
    await client().auth.login({ email: 'a@b.dev', password: 'pw', lat: 51.5 });
    assert.deepEqual(JSON.parse(recorded[0]!.body), {
      email: 'a@b.dev',
      password: 'pw',
    });

    reset();
    responses.push({ status: 200, body: '{"token":"t","expires_at":1}' });
    await client().auth.login({
      email: 'a@b.dev',
      password: 'pw',
      lat: 51.5,
      lng: -0.1,
    });
    assert.deepEqual(JSON.parse(recorded[0]!.body), {
      email: 'a@b.dev',
      password: 'pw',
      lat: 51.5,
      lng: -0.1,
    });
  });

  test('logout clears the token even when the server says false', async () => {
    reset();
    responses.push({ status: 200, body: '{"success":false}' });
    const c = client({ token: 'tok-1' });
    await c.auth.logout();
    assert.equal(c.getToken(), undefined);
  });
});

describe('geo', () => {
  test('nearby sends snake_case query params and maps the response', async () => {
    reset();
    responses.push({
      status: 200,
      body: '{"users":[{"user_id":"u-2","lat":1,"lng":2,"distance_m":42.5,"safety_score":1500}]}',
    });
    const { users } = await client({ token: 't' }).geo.nearby({
      lat: 1,
      lng: 2,
      radiusM: 500,
      limit: 10,
    });

    const url = new URL(baseUrl + recorded[0]!.url);
    assert.equal(url.pathname, '/v1/geo/nearby');
    assert.equal(url.searchParams.get('radius_m'), '500');
    assert.equal(url.searchParams.get('limit'), '10');
    // Omitted optional params must not appear as the string "undefined".
    assert.equal(url.searchParams.has('role'), false);

    assert.equal(users[0]?.userId, 'u-2');
    assert.equal(users[0]?.distanceM, 42.5);
    assert.equal(users[0]?.safetyScore, 1500);
  });

  test('scoreRoute maps nested route candidates both ways', async () => {
    reset();
    responses.push({
      status: 200,
      body: '{"best_route_id":"r-1","score":1600,"all_scores":[{"route_id":"r-1","score":1600}]}',
    });
    const out = await client({ token: 't' }).geo.scoreRoute([
      { routeId: 'r-1', points: [{ lat: 1, lng: 2 }, { lat: 3, lng: 4 }] },
    ]);

    assert.deepEqual(JSON.parse(recorded[0]!.body), {
      candidates: [
        { route_id: 'r-1', points: [{ lat: 1, lng: 2 }, { lat: 3, lng: 4 }] },
      ],
    });
    assert.equal(out.bestRouteId, 'r-1');
    assert.equal(out.allScores[0]?.routeId, 'r-1');
  });

  test('deleteGeofence url-encodes the id', async () => {
    reset();
    responses.push({ status: 200, body: '{"deleted":true}' });
    // A caller passing something odd must not be able to escape the path.
    await client({ token: 't' }).geo.deleteGeofence('a/../b');
    assert.equal(recorded[0]?.url, '/v1/geo/geofences/a%2F..%2Fb');
  });
});

describe('payments', () => {
  test('createTransaction always sends an Idempotency-Key', async () => {
    reset();
    responses.push({
      status: 201,
      body: '{"transaction_id":"tx-1","status":"pending"}',
    });
    const out = await client({ token: 't' }).payments.createTransaction({
      toUserId: 'u-2',
      amountCents: 500,
    });

    const key = recorded[0]?.headers['idempotency-key'];
    assert.ok(typeof key === 'string' && key.length > 0, 'key must be sent');
    assert.equal(out.transactionId, 'tx-1');
    // No user id in the body: identity comes from the token.
    assert.deepEqual(JSON.parse(recorded[0]!.body), {
      to_user_id: 'u-2',
      amount_cents: 500,
    });
  });

  test('deposit sends an idempotency key and maps the balance back', async () => {
    reset();
    responses.push({
      status: 201,
      body: '{"transaction_id":"tx-9","status":"settled","balance_cents":10000}',
    });
    const out = await client({ token: 't' }).payments.deposit({ amountCents: 10_000 });

    assert.equal(recorded[0]?.url, '/v1/payments/deposits');
    assert.ok(typeof recorded[0]?.headers['idempotency-key'] === 'string');
    // No toUserId: a caller can only ever top up themselves.
    assert.deepEqual(JSON.parse(recorded[0]!.body), { amount_cents: 10_000 });
    assert.equal(out.status, 'settled');
    assert.equal(out.balanceCents, 10_000);
  });

  test('a supplied idempotency key is used verbatim', async () => {
    reset();
    responses.push({ status: 201, body: '{"transaction_id":"t","status":"pending"}' });
    await client({ token: 't' }).payments.createTransaction({
      toUserId: 'u-2',
      amountCents: 500,
      idempotencyKey: 'mine-123',
    });
    assert.equal(recorded[0]?.headers['idempotency-key'], 'mine-123');
  });
});

describe('errors', () => {
  test('the error envelope becomes an AtlasError with a stable code', async () => {
    reset();
    responses.push({
      status: 422,
      body: '{"error":{"code":"failed_precondition","message":"insufficient funds"}}',
    });
    await assert.rejects(
      () =>
        client({ token: 't' }).payments.createTransaction({
          toUserId: 'u-2',
          amountCents: 1,
        }),
      (err: unknown) => {
        assert.ok(err instanceof AtlasError);
        assert.equal(err.code, 'failed_precondition');
        assert.equal(err.status, 422);
        assert.equal(err.message, 'insufficient funds');
        assert.equal(err.retryable, false);
        return true;
      },
    );
  });

  test('a non-envelope error body still yields a useful error', async () => {
    reset();
    // What a load balancer returns when it never reached the gateway.
    responses.push({
      status: 502,
      body: '<html>502 Bad Gateway</html>',
      contentType: 'text/html',
    });
    await assert.rejects(
      () => client({ token: 't' }).payments.wallet(),
      (err: unknown) => {
        assert.ok(err instanceof AtlasError);
        assert.equal(err.status, 502);
        assert.ok(err.message.includes('502'));
        return true;
      },
    );
  });

  test('connection failures are a distinct type', async () => {
    const c = new AtlasClient({
      // Port nothing listens on.
      baseUrl: 'http://127.0.0.1:59987',
      projectKey: TEST_PROJECT_KEY,
      maxRetries: 0,
      timeoutMs: 1_000,
    });
    await assert.rejects(
      () => c.auth.register({ email: 'a@b.dev', password: 'pw' }),
      (err: unknown) => {
        assert.ok(err instanceof AtlasConnectionError);
        return true;
      },
    );
  });
});

describe('retries', () => {
  test('GET retries a 503 and then succeeds', async () => {
    reset();
    responses.push({
      status: 503,
      body: '{"error":{"code":"unavailable","message":"backend down"}}',
    });
    responses.push({ status: 200, body: '{"balance_cents":100,"currency":"USD"}' });

    const out = await client({ token: 't', maxRetries: 2 }).payments.wallet();
    assert.equal(out.balanceCents, 100);
    assert.equal(recorded.length, 2, 'should have retried once');
  });

  test('a 4xx is not retried', async () => {
    reset();
    responses.push({
      status: 401,
      body: '{"error":{"code":"unauthenticated","message":"invalid token"}}',
    });
    await assert.rejects(() =>
      client({ token: 'bad', maxRetries: 2 }).payments.wallet(),
    );
    assert.equal(recorded.length, 1, 'client errors must not be retried');
  });

  test('POST with an idempotency key is retried', async () => {
    reset();
    responses.push({
      status: 503,
      body: '{"error":{"code":"unavailable","message":"down"}}',
    });
    responses.push({
      status: 201,
      body: '{"transaction_id":"tx-1","status":"pending"}',
    });

    const out = await client({ token: 't', maxRetries: 2 }).payments.createTransaction({
      toUserId: 'u-2',
      amountCents: 500,
    });
    assert.equal(out.transactionId, 'tx-1');
    assert.equal(recorded.length, 2);
    // Critically, the same key on both attempts — otherwise the retry is a
    // second distinct transaction and the payer is charged twice.
    assert.equal(
      recorded[0]?.headers['idempotency-key'],
      recorded[1]?.headers['idempotency-key'],
    );
  });

  test('POST without an idempotency key is never retried', async () => {
    reset();
    responses.push({
      status: 503,
      body: '{"error":{"code":"unavailable","message":"down"}}',
    });
    responses.push({ status: 200, body: '{"ok":true}' });

    // updateLocation carries no key, so repeating it is not this layer's
    // call to make.
    await assert.rejects(() =>
      client({ token: 't', maxRetries: 2 }).geo.updateLocation({ lat: 1, lng: 2 }),
    );
    assert.equal(recorded.length, 1);
  });
});

describe('project key', () => {
  test('is sent on every request, authenticated or not', async () => {
    reset();
    await client().auth.register({ email: 'a@b.dev', password: 'pw' });
    reset();
    await client({ token: 't' }).payments.wallet();

    // Both calls above recorded one request each; check the second, which
    // also carries a bearer token, to prove the two headers coexist.
    assert.equal(recorded[0]?.headers['x-atlas-key'], TEST_PROJECT_KEY);
    assert.equal(recorded[0]?.headers['authorization'], 'Bearer t');
  });

  test('a missing project key fails at construction, not at call time', () => {
    assert.throws(
      // Cast because the type already forbids this; the runtime check is
      // for JavaScript callers and for a key read from an unset env var.
      () => new AtlasClient({ baseUrl, projectKey: '' } as never),
      /projectKey/,
    );
  });

  test('caller headers cannot override the project key', async () => {
    // Otherwise a stray `headers` option would silently send requests as
    // some other project, which is the sort of thing that only shows up
    // in production.
    reset();
    const c = new AtlasClient({
      baseUrl,
      projectKey: TEST_PROJECT_KEY,
      maxRetries: 0,
      headers: { 'X-Atlas-Key': 'atl_live_ffffffffffffffffffffffffffffffff' },
    });
    await c.auth.register({ email: 'a@b.dev', password: 'pw' });
    assert.equal(recorded[0]?.headers['x-atlas-key'], TEST_PROJECT_KEY);
  });
});

File: sdks/dart/test/client_test.dart
// Tests against a real HTTP server on a real port.
//
// Not a mocked transport: the point is to exercise the actual request path
// — headers, query encoding, retries, JSON handling — the way it runs in
// production. A mock only proves the SDK can call a mock.

import 'dart:convert';
import 'dart:io';

import 'package:atlas_sdk/atlas_sdk.dart';
import 'package:test/test.dart';

const testKey = 'atl_test_0123456789abcdef0123456789abcdef';

/// What the fake gateway saw, so tests can assert on the wire.
class Recorded {
  Recorded(this.method, this.path, this.query, this.key, this.authorization,
      this.body);

  final String method;
  final String path;
  final Map<String, String> query;
  final String? key;
  final String? authorization;
  final String body;
}

class FakeGateway {
  FakeGateway._(this._server);

  static Future<FakeGateway> start() async {
    final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
    final gateway = FakeGateway._(server);
    gateway._serve();
    return gateway;
  }

  final HttpServer _server;
  final List<Recorded> requests = [];

  /// Fail this many times before succeeding, for the retry tests.
  int failTimes = 0;
  int attempts = 0;

  String get baseUrl => 'http://127.0.0.1:${_server.port}';

  Future<void> close() => _server.close(force: true);

  void _serve() {
    _server.listen((request) async {
      final body = await utf8.decoder.bind(request).join();
      requests.add(Recorded(
        request.method,
        request.uri.path,
        request.uri.queryParameters,
        request.headers.value('x-atlas-key'),
        request.headers.value('authorization'),
        body,
      ));

      request.response.headers.contentType = ContentType.json;

      // Endpoints that participate in the retry tests.
      if (request.uri.path == '/v1/payments/wallet' ||
          request.uri.path == '/v1/payments/deposits') {
        final seen = attempts++;
        if (seen < failTimes) {
          request.response.statusCode = 503;
          request.response.write(jsonEncode({
            'error': {'code': 'unavailable', 'message': 'try again'}
          }));
          await request.response.close();
          return;
        }
      }

      final payload = switch (request.uri.path) {
        '/v1/auth/register' => {'user_id': 'u-1'},
        '/v1/auth/login' => {'token': 'tok-abc', 'expires_at': 1787000000},
        '/v1/auth/logout' => {'success': true},
        '/v1/auth/me' => {
            'user_id': 'u-1',
            'session_id': 's-1',
            'last_lat': 0.0,
            'last_lng': 0.0,
            'issued_at': 1,
            'expires_at': 2,
            'email': 'a@b.dev',
            'email_verified_at': null,
            'created_at': 1,
          },
        '/v1/auth/password-reset' => {'accepted': true},
        '/v1/auth/password-reset/confirm' => {'user_id': 'u-1'},
        '/v1/auth/email/verify/confirm' => {'user_id': 'u-1'},
        '/v1/geo/locations' => {'ok': true},
        '/v1/geo/nearby' => {
            'users': [
              {
                'user_id': 'u-2',
                'lat': 1.0,
                'lng': 2.0,
                'distance_m': 3.0,
                'safety_score': 1500.0,
                'safety_vote_count': 0,
              }
            ]
          },
        '/v1/geo/safety/votes' => {'safety_score': 1642.8, 'vote_count': 2},
        '/v1/payments/wallet' => {'balance_cents': 5000, 'currency': 'USD'},
        '/v1/payments/deposits' => {
            'transaction_id': 't-1',
            'status': 'settled',
            'balance_cents': 5000,
          },
        _ => null,
      };

      if (payload == null) {
        // Every unrecognised path is a 404 with the real envelope, so the
        // error mapping is exercised by the geofence delete test.
        request.response.statusCode = 404;
        request.response.write(jsonEncode({
          'error': {'code': 'not_found', 'message': 'geofence not found'}
        }));
      } else {
        request.response.write(jsonEncode(payload));
      }
      await request.response.close();
    });
  }
}

void main() {
  late FakeGateway gateway;

  setUp(() async {
    gateway = await FakeGateway.start();
  });

  tearDown(() async {
    await gateway.close();
  });

  AtlasClient client({int maxRetries = 0, String? token}) => AtlasClient(
        baseUrl: gateway.baseUrl,
        projectKey: testKey,
        token: token,
        maxRetries: maxRetries,
        timeout: const Duration(seconds: 5),
      );

  group('credentials', () {
    test('a missing project key throws at construction', () {
      // Not at call time: a missing key is a configuration mistake, and
      // the stack trace should point at the line that has to change.
      expect(
        () => AtlasClient(baseUrl: 'http://127.0.0.1:1', projectKey: ''),
        throwsA(isA<ArgumentError>()),
      );
    });

    test('the project key is sent on every call, including register', () async {
      final c = client();
      await c.auth.register(email: 'a@b.dev', password: 'hunter2!');
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');
      await c.auth.me();
      c.close();

      expect(gateway.requests, hasLength(3));
      for (final r in gateway.requests) {
        expect(r.key, testKey, reason: 'missing key on ${r.path}');
      }
      // Register and login carry no bearer; /me does.
      expect(gateway.requests[0].authorization, isNull);
      expect(gateway.requests[1].authorization, isNull);
      expect(gateway.requests[2].authorization, 'Bearer tok-abc');
    });

    test('login stores the token, logout forgets it', () async {
      final c = client();
      expect(c.token, isNull);
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');
      expect(c.token, 'tok-abc');
      await c.auth.logout();
      expect(c.token, isNull);
      c.close();
    });

    test('a completed reset clears the token', () async {
      // The server just revoked every session; holding the dead token
      // would only produce confusing 403s on the next call.
      final c = client();
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');
      await c.auth.resetPassword(token: 'f' * 64, newPassword: 'a-new-one');
      expect(c.token, isNull);
      c.close();
    });

    test('verifying an email leaves the session alone', () async {
      final c = client();
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');
      await c.auth.verifyEmail(token: 'a' * 64);
      expect(c.token, 'tok-abc');
      c.close();
    });

    test('toString never contains the project key or token', () async {
      // The default would print the key the first time anything logs a
      // client, and it would land in a crash report.
      final c = client();
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');
      expect(c.toString(), isNot(contains(testKey)));
      expect(c.toString(), isNot(contains('tok-abc')));
      expect(c.toString(), contains('authenticated: true'));
      c.close();
    });
  });

  group('identity', () {
    test('no request body carries a user id or project id', () async {
      final c = client();
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');
      await c.geo.castSafetyVote(lat: 1, lng: 2, verdict: Verdict.unsafe);
      await c.geo.updateLocation(lat: 1, lng: 2);
      c.close();

      for (final r in gateway.requests) {
        expect(r.body, isNot(contains('user_id')), reason: r.path);
        expect(r.body, isNot(contains('project_id')), reason: r.path);
      }
    });
  });

  group('transport', () {
    test('a GET is retried and eventually succeeds', () async {
      gateway.failTimes = 2;
      final c = client(maxRetries: 2);
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');

      final wallet = await c.payments.wallet();
      expect(wallet.balanceCents, 5000);
      expect(gateway.attempts, 3, reason: 'two failures, one success');
      c.close();
    });

    test('a deposit is retried and reuses its idempotency key', () async {
      // The key is what makes replaying a POST safe on the server side. A
      // fresh key per retry would double-charge.
      gateway.failTimes = 1;
      final c = client(maxRetries: 2);
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');

      final deposit = await c.payments.deposit(amountCents: 5000);
      expect(deposit.balanceCents, 5000);

      final bodies = gateway.requests
          .where((r) => r.path == '/v1/payments/deposits')
          .map((r) => r.body)
          .toList();
      expect(bodies, hasLength(2));
      expect(bodies[0], bodies[1], reason: 'the retry must reuse the key');
      c.close();
    });

    test('query parameters are encoded', () async {
      final c = client();
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');
      await c.geo.nearby(lat: 51.5074, lng: -0.1278, radiusM: 500);

      final q =
          gateway.requests.firstWhere((r) => r.path == '/v1/geo/nearby').query;
      expect(q['lat'], '51.5074');
      expect(q['lng'], '-0.1278');
      expect(q['radius_m'], '500.0');
      c.close();
    });
  });

  group('errors', () {
    test('an error envelope becomes a typed code', () async {
      final c = client();
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');

      await expectLater(
        c.geo.deleteGeofence('someone-elses'),
        throwsA(isA<AtlasError>()
            .having((e) => e.code, 'code', AtlasErrorCode.notFound)
            .having((e) => e.isRetryable, 'isRetryable', isFalse)),
      );
      c.close();
    });

    test('an unreachable gateway is a connection error, not an API error',
        () async {
      // The distinction matters: "the service rejected this" and "we could
      // not ask" need different handling, and conflating them produces
      // both spurious alerts and missed outages.
      final c = AtlasClient(
        // A port nothing listens on.
        baseUrl: 'http://127.0.0.1:59987',
        projectKey: testKey,
        maxRetries: 0,
        timeout: const Duration(milliseconds: 500),
      );

      await expectLater(
        c.auth.register(email: 'a@b.dev', password: 'hunter2!'),
        throwsA(isA<AtlasConnectionError>()),
      );
      c.close();
    });

    test('scoring no routes fails before the request', () async {
      final c = client();
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');
      final before = gateway.requests.length;

      expect(() => c.geo.scoreRoute([]), throwsA(isA<ArgumentError>()));
      expect(gateway.requests.length, before,
          reason: 'an obviously invalid request should not cost a round trip');
      c.close();
    });
  });

  group('types', () {
    test('an unverified address is null, not zero', () async {
      // 0 is a real timestamp; a caller doing date maths on it renders
      // 1970.
      final c = client();
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');

      final me = await c.auth.me();
      expect(me.emailVerifiedAt, isNull);
      expect(me.isEmailVerified, isFalse);
      expect(me.email, 'a@b.dev');
      c.close();
    });

    test('a neutral score arrives with its evidence', () async {
      final c = client();
      await c.auth.login(email: 'a@b.dev', password: 'hunter2!');

      final users = await c.geo.nearby(lat: 1, lng: 2, radiusM: 100);
      expect(users.single.safetyScore, 1500.0);
      // Zero voters means the score is a default, not a measurement.
      expect(users.single.safetyVoteCount, 0);
      expect(users.single.hasSafetyData, isFalse);
      c.close();
    });
  });
}
